pkg/cache/cache_test.go

```diff
@@ -255,21 +262,53 @@ func CacheTest(createCacheFunc func(config *rest.Config, opts cache.Options) (ca
 					}
 				})
 
-				It("should deep copy the object unless told otherwise", func() {
-					By("retrieving a specific pod from the cache")
-					out := &corev1.Pod{}
-					podKey := client.ObjectKey{Name: "test-pod-2", Namespace: testNamespaceTwo}
-					Expect(informerCache.Get(context.Background(), podKey, out)).To(Succeed())
+				if opts.UnsafeDisableCacheDeepCopy == nil {
+					It("should deep copy the object unless told otherwise", func() {
+						By("retrieving a specific pod from the cache")
+						out := &corev1.Pod{}
+						podKey := client.ObjectKey{Name: "test-pod-2", Namespace: testNamespaceTwo}
+						Expect(informerCache.Get(context.Background(), podKey, out)).To(Succeed())
 
-					By("verifying the retrieved pod is equal to a known pod")
-					Expect(out).To(Equal(knownPod2))
+						By("verifying the retrieved pod is equal to a known pod")
+						Expect(out).To(Equal(knownPod2))
 
-					By("altering a field in the retrieved pod")
-					*out.Spec.ActiveDeadlineSeconds = 4
+						By("altering a field in the retrieved pod")
+						*out.Spec.ActiveDeadlineSeconds = 4
 
-					By("verifying the pods are no longer equal")
-					Expect(out).NotTo(Equal(knownPod2))
-				})
+						By("verifying the pods are no longer equal")
+						Expect(out).NotTo(Equal(knownPod2))
+					})
+				} else {
+					It("should not deep copy the object if UnsafeDisableCacheDeepCopy is enabled", func() {
+						By("getting a specific pod from the cache twice")
+						podKey := client.ObjectKey{Name: "test-pod-2", Namespace: testNamespaceTwo}
+						out1 := &corev1.Pod{}
+						Expect(informerCache.Get(context.Background(), podKey, out1)).To(Succeed())
+						out2 := &corev1.Pod{}
+						Expect(informerCache.Get(context.Background(), podKey, out2)).To(Succeed())
+
+						By("verifying the pointer fields in pod have the same addresses")
+						Expect(out1).To(Equal(out2))
+						Expect(reflect.ValueOf(out1.Labels).Pointer()).To(BeIdenticalTo(reflect.ValueOf(out2.Labels).Pointer()))
+
```

Comment:
Can we add another expectation that verifies the GVK is set on the returned objects after `Get()` and `List()` when deep copying is disabled?

Reply:
When deep copying is disabled, we could not ensure the GVK is set on the returned objects.